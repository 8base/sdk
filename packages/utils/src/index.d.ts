
export type FieldSchema = {
  id: string,
  name: string,
  displayName?: string,
  description?: string,
  fieldType: string,
  fieldTypeAttributes: Object,
  isSystem: boolean,
  isList: boolean,
  isMeta: boolean,
  isRequired: boolean,
  isExternal: boolean,
  isUnique?: boolean,
  defaultValue: any,
  relation: {
    id: string,
    relationTableName: string,
    relationFieldName: string,
    refFieldName: string,
    refFieldDisplayName: string,
    refFieldIsList: boolean,
    refFieldIsRequired: boolean,
    refTable: {
      id: string,
    },
  },
};

export type TableSchema = {
  id: string,
  name: string,
  displayName?: string,
  isSystem?: boolean,
  fields: Array<FieldSchema>,
};

export type QueryGeneratorConfig = {
  deep?: number,
  withMeta?: boolean,
  relationItemsCount?: number,
  withResultData?: boolean,
  includeColumns?: null | string[],
};

type CreateQueryColumnsListConfig = {
  flatten?: boolean,
} & QueryGeneratorConfig


export type Schema = Array<TableSchema>;

export function createQueryColumnsList(
  tablesList: TableSchema[],
  tableName: string,
  config?: CreateQueryColumnsListConfig,
): Array<{ name: string, title: string, meta: Object }>;


export const FIELD_TYPE: {
  DATE: 'DATE',
  FILE: 'FILE',
  ID: 'ID',
  NUMBER: 'NUMBER',
  RELATION: 'RELATION',
  SWITCH: 'SWITCH',
  TEXT: 'TEXT',
  SMART: 'SMART',
  JSON: 'JSON',
};

export const SWITCH_FORMATS: {
  ON_OFF: 'ON_OFF',
  YES_NO: 'YES_NO',
  TRUE_FALSE: 'TRUE_FALSE',
  ACTIVE_INACTIVE: 'ACTIVE_INACTIVE',
  HIGH_LOW: 'HIGH_LOW',
  CUSTOM: 'CUSTOM',
};

export const SWITCH_VALUES: {
  ON_OFF: {
    true: 'On',
    false: 'Off',
  },
  YES_NO: {
    true: 'Yes',
    false: 'No',
  },
  TRUE_FALSE: {
    true: 'True',
    false: 'False',
  },
  ACTIVE_INACTIVE: {
    true: 'Active',
    false: 'Inactive',
  },
  HIGH_LOW: {
    true: 'High',
    false: 'Low',
  },
};

export const TEXT_FORMATS: {
  UNFORMATTED: 'UNFORMATTED',
  NAME: 'NAME',
  SSN: 'SSN',
  EIN: 'EIN',
  PHONE: 'PHONE',
  EMAIL: 'EMAIL',
};

export const NUMBER_FORMATS: {
  NUMBER: 'NUMBER',
  CURRENCY: 'CURRENCY',
  PERCENTAGE: 'PERCENTAGE',
  FRACTION: 'FRACTION',
  SCIENTIFIC: 'SCIENTIFIC',
};

export const FILE_FORMATS: {
  FILE: 'FILE',
  IMAGE: 'IMAGE',
};

export const SMART_FORMATS: {
  ADDRESS: 'ADDRESS',
  PHONE: 'PHONE',
};

export const DATE_FORMATS: {
  DATE: 'DATE',
  DATETIME: 'DATETIME',
};

export const MUTATION_TYPE: {
  CREATE: 'CREATE',
  UPDATE: 'UPDATE',
};

export const SYSTEM_TABLES: {
  USERS: 'Users',
  FILES: 'Files',
  SETTINGS: 'Settings',
  ROLES: 'Roles',
  INVITATIONS: 'Invitations',
  PERMISSIONS: 'Permissions',
  API_TOKENS: 'ApiTokens',
  ENVIRONMENT_VARIABLES: 'EnvironmentVariables',
};

export const MUTATION_FILE_FIELDS: ['fileId', 'public', 'filename', 'meta', 'mods'];

export type QueryGeneratorData = {
  tablesList: TableSchema[],
  tableName: string,
  appName?: string,
};

export const tablesListSelectors: {
  [key: string]: Function
}

export const tableSelectors: {
  [key: string]: Function
}

export const tableFieldSelectors: { [key: string]: Function };

export const throwIfMissingRequiredOption: Function;

export type AuthState = {
  email?: string,
  userId?: string,
  workspaceId?: string,
  refreshToken?: string,
  token?: string,
}

export type AuthData = {
  isEmailVerified: boolean,
  idToken: string,
  email: string,
  idTokenPayload: Object,
  state: any,
};

export interface AuthClient {
  getAuthState(): Promise<AuthState>,
  setAuthState(state: AuthState): Promise<void>,
  purgeAuthState(): Promise<void>,
  checkIsAuthorized(): Promise<boolean>,
}

export interface Authorizable {
  authorize(options?: Object): Promise<AuthData>,
  renewToken(options?: Object): Promise<AuthData>,
  changePassword(): Promise<{ email: string }>,
  logout(options?: Object): Promise<void>,
}
