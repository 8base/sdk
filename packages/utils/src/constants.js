//@flow

export const FIELD_TYPE = {
  DATE: 'DATE',
  FILE: 'FILE',
  ID: 'ID',
  NUMBER: 'NUMBER',
  RELATION: 'RELATION',
  SWITCH: 'SWITCH',
  TEXT: 'TEXT',
  SMART: 'SMART',
};

export const SWITCH_FORMATS = {
  ON_OFF: 'ON_OFF',
  YES_NO: 'YES_NO',
  TRUE_FALSE: 'TRUE_FALSE',
  ACTIVE_INACTIVE: 'ACTIVE_INACTIVE',
  HIGH_LOW: 'HIGH_LOW',
  SMART: 'SMART',
};

export const TEXT_FORMATS = {
  UNFORMATTED: 'UNFORMATTED',
  NAME: 'NAME',
  ADDRESS: 'ADDRESS',
  SSN: 'SSN',
  EIN: 'EIN',
  PHONE: 'PHONE',
  EMAIL: 'EMAIL',
};

export const NUMBER_FORMATS = {
  NUMBER: 'NUMBER',
  CURRENCY: 'CURRENCY',
  PERCENTAGE: 'PERCENTAGE',
  FRACTION: 'FRACTION',
  SCIENTIFIC: 'SCIENTIFIC',
};

export const FILE_FORMATS = {
  FILE: 'FILE',
  IMAGE: 'IMAGE',
};

export const DATE_FORMATS = {
  DATE: 'DATE',
  DATETIME: 'DATE/TIME',
};

export const MUTATION_TYPE = {
  CREATE: 'CREATE',
  UPDATE: 'UPDATE',
};
