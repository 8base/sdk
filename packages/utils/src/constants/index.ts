// @flow

export * from './schemaConstants';
export * from './countryCodes';
