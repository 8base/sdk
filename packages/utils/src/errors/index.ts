export { SDKError } from './SDKError';
export { ERROR_CODES } from './codes';

