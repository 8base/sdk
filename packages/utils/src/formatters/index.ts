export { formatDataForMutation } from './formatDataForMutation';
export { formatDataAfterQuery } from './formatDataAfterQuery';
