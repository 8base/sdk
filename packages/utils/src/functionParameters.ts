// @flow
import * as R from 'ramda';

const hasOption: (string | Array<string>) => ({}) => boolean = R.ifElse(
  R.is(String),
  R.has,
  R.hasPath,
);

const stringifyPath: (string | Array<string>) => string = R.ifElse(
  R.is(String),
  (path) => path,
  R.join('.'),
);

export const throwIfMissing = (paramName: string): void => {
  throw new Error(`Missing parameter: ${paramName}`);
};

export const throwIfMissingRequiredOption = (
  requiredOptionPaths: Array<Array<string> | string>,
  options: {},
): void => {
  requiredOptionPaths.forEach(optionPath => {
    const isMissing = !hasOption(optionPath)(options);

    if (isMissing) {
      throw new Error(`Missing option: ${stringifyPath(optionPath)}`);
    }
  });
};

