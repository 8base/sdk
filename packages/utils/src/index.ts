import * as tableSelectors from './selectors/tableSelectors';

export * from './constants';
export * from './formatters';
export * from './selectors';
export * from './verifiers';
export * from './queryGenerators';
export * from './functionParameters';
export * from './errors';

export { tableSelectors };
