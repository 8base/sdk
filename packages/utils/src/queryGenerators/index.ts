// @flow

export * from './queryGenerators';
export { createQueryColumnsList } from './createQueryColumnsList';
