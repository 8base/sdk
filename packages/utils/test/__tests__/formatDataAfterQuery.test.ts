
import { formatDataAfterQuery } from '../../src';
import { SCHEMA } from '../__fixtures__';

describe('As developer, I can format for create mutation,', () => {
  it('Compelex data.', () => {
    const data = {
      meta: 'meta',
      address: {
        street1: 'Pamelia Quall',
        street2: 'Lasonya Friedly',
        zip: 'Timothy Ingleton',
        city: 'Kenia Urhahn',
        state: 'Scottie Swailes',
      },
      scalar: 'Scalar Value',
      scalarList: [
        'Scalar List Value',
      ],
      relation: null,
      relationList: {
        items: [{
          id: 'id-1',
          scalar: 'Relation List Scalar Value',
          scalarList: [
            'Relation List Scalar List Value',
          ],
          nestedRelation: {
            id: 'nestedId',
            scalar: 'Nested Scalar Value',
            nonNestedRelationField: 'should be removed',
          },
        }, {
          id: 'id-2',
          scalar: 'Relation List Scalar Value',
          scalarList: [
            'Relation List Scalar List Value',
          ],
        }],
      },
      fileList: {
        items: [{
          id: 'cjnodqoni000501qnkcga0qra',
          fileId: '9SNTzTfTAWMFarHrohA6',
          filename: '9217163.png',
          downloadUrl: 'https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI5U05UelRmVEFXTUZhckhyb2hBNiJ9,s:d019cfe489a72037a9fd57bfc008664b1ec5bd7d4e24f0979aa067d29d54c305/9SNTzTfTAWMFarHrohA6',
          shareUrl: 'https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqoni000501qnkcga0qra',
        }, {
          id: 'cjnodqop0000601qnytfd866c',
          fileId: '5AXQF5SaQRmgrxTJCMqQ',
          filename: '9217163.png',
          downloadUrl: 'https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI1QVhRRjVTYVFSbWdyeFRKQ01xUSJ9,s:1ee9fe13f9a855acd1aad3330a7612b4c6849a7de4f9d45d5b831e596ca26115/5AXQF5SaQRmgrxTJCMqQ',
          shareUrl: 'https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqop0000601qnytfd866c',
        }],
      },
      json: {
        somePropArray: [
          'someValue1',
          'someValue2',
          'someValue3',
        ],
      },
      jsonList: [
        { someProp: 'someValue' },
        {
          somePropArray: [
            'someValue1',
            'someValue2',
          ],
        },
      ],
    };

    expect(formatDataAfterQuery('tableSchema', data, SCHEMA)).toMatchSnapshot();
  });

  it('Compelex data with formatRelationToIds option.', () => {
    const data = {
      meta: 'meta',
      address: {
        street1: 'Pamelia Quall',
        street2: 'Lasonya Friedly',
        zip: 'Timothy Ingleton',
        city: 'Kenia Urhahn',
        state: 'Scottie Swailes',
      },
      scalar: 'Scalar Value',
      scalarList: [
        'Scalar List Value',
      ],
      relation: null,
      relationList: {
        items: [{
          id: 'id-1',
          scalar: 'Relation List Scalar Value',
          scalarList: [
            'Relation List Scalar List Value',
          ],
          nestedRelation: {
            id: 'nestedId',
            scalar: 'Nested Scalar Value',
            nonNestedRelationField: 'should be removed',
          },
        }, {
          id: 'id-2',
          scalar: 'Relation List Scalar Value',
          scalarList: [
            'Relation List Scalar List Value',
          ],
        }],
      },
      fileList: {
        items: [{
          id: 'cjnodqoni000501qnkcga0qra',
          fileId: '9SNTzTfTAWMFarHrohA6',
          filename: '9217163.png',
          downloadUrl: 'https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI5U05UelRmVEFXTUZhckhyb2hBNiJ9,s:d019cfe489a72037a9fd57bfc008664b1ec5bd7d4e24f0979aa067d29d54c305/9SNTzTfTAWMFarHrohA6',
          shareUrl: 'https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqoni000501qnkcga0qra',
        }, {
          id: 'cjnodqop0000601qnytfd866c',
          fileId: '5AXQF5SaQRmgrxTJCMqQ',
          filename: '9217163.png',
          downloadUrl: 'https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI1QVhRRjVTYVFSbWdyeFRKQ01xUSJ9,s:1ee9fe13f9a855acd1aad3330a7612b4c6849a7de4f9d45d5b831e596ca26115/5AXQF5SaQRmgrxTJCMqQ',
          shareUrl: 'https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqop0000601qnytfd866c',
        }],
      },
      json: {
        somePropArray: [
          'someValue1',
          'someValue2',
          'someValue3',
        ],
      },
      jsonList: [
        { someProp: 'someValue' },
        {
          somePropArray: [
            'someValue1',
            'someValue2',
          ],
        },
      ],
    };

    expect(formatDataAfterQuery('tableSchema', data, SCHEMA, { formatRelationToIds: true })).toMatchSnapshot();
  });
});
