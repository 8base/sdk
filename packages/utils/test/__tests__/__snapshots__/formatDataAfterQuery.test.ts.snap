// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`As developer, I can format for create mutation, Compelex data with formatRelationToIds option. 1`] = `
Object {
  "address": Object {
    "city": "Kenia Urhahn",
    "state": "Scottie Swailes",
    "street1": "Pamelia Quall",
    "street2": "Lasonya Friedly",
    "zip": "Timothy Ingleton",
  },
  "fileList": Array [
    Object {
      "downloadUrl": "https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI5U05UelRmVEFXTUZhckhyb2hBNiJ9,s:d019cfe489a72037a9fd57bfc008664b1ec5bd7d4e24f0979aa067d29d54c305/9SNTzTfTAWMFarHrohA6",
      "fileId": "9SNTzTfTAWMFarHrohA6",
      "filename": "9217163.png",
      "id": "cjnodqoni000501qnkcga0qra",
      "shareUrl": "https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqoni000501qnkcga0qra",
    },
    Object {
      "downloadUrl": "https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI1QVhRRjVTYVFSbWdyeFRKQ01xUSJ9,s:1ee9fe13f9a855acd1aad3330a7612b4c6849a7de4f9d45d5b831e596ca26115/5AXQF5SaQRmgrxTJCMqQ",
      "fileId": "5AXQF5SaQRmgrxTJCMqQ",
      "filename": "9217163.png",
      "id": "cjnodqop0000601qnytfd866c",
      "shareUrl": "https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqop0000601qnytfd866c",
    },
  ],
  "json": Object {
    "somePropArray": Array [
      "someValue1",
      "someValue2",
      "someValue3",
    ],
  },
  "jsonList": Array [
    Object {
      "someProp": "someValue",
    },
    Object {
      "somePropArray": Array [
        "someValue1",
        "someValue2",
      ],
    },
  ],
  "relation": null,
  "relationList": Array [
    "id-1",
    "id-2",
  ],
  "scalar": "Scalar Value",
  "scalarList": Array [
    "Scalar List Value",
  ],
}
`;

exports[`As developer, I can format for create mutation, Compelex data. 1`] = `
Object {
  "address": Object {
    "city": "Kenia Urhahn",
    "state": "Scottie Swailes",
    "street1": "Pamelia Quall",
    "street2": "Lasonya Friedly",
    "zip": "Timothy Ingleton",
  },
  "fileList": Array [
    Object {
      "downloadUrl": "https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI5U05UelRmVEFXTUZhckhyb2hBNiJ9,s:d019cfe489a72037a9fd57bfc008664b1ec5bd7d4e24f0979aa067d29d54c305/9SNTzTfTAWMFarHrohA6",
      "fileId": "9SNTzTfTAWMFarHrohA6",
      "filename": "9217163.png",
      "id": "cjnodqoni000501qnkcga0qra",
      "shareUrl": "https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqoni000501qnkcga0qra",
    },
    Object {
      "downloadUrl": "https://cdn.filestackcontent.com/security=p:eyJjYWxsIjpbInJlYWQiXSwiZXhwaXJ5IjoxNTQwNDU5ODAxODYzLCJoYW5kbGUiOiI1QVhRRjVTYVFSbWdyeFRKQ01xUSJ9,s:1ee9fe13f9a855acd1aad3330a7612b4c6849a7de4f9d45d5b831e596ca26115/5AXQF5SaQRmgrxTJCMqQ",
      "fileId": "5AXQF5SaQRmgrxTJCMqQ",
      "filename": "9217163.png",
      "id": "cjnodqop0000601qnytfd866c",
      "shareUrl": "https://prestaging-api.8basedev.com/file/download/cjnod10tj000001pjf7an05k2_Master/cjnodqop0000601qnytfd866c",
    },
  ],
  "json": Object {
    "somePropArray": Array [
      "someValue1",
      "someValue2",
      "someValue3",
    ],
  },
  "jsonList": Array [
    Object {
      "someProp": "someValue",
    },
    Object {
      "somePropArray": Array [
        "someValue1",
        "someValue2",
      ],
    },
  ],
  "relation": null,
  "relationList": Array [
    Object {
      "id": "id-1",
      "nestedRelation": Object {
        "id": "nestedId",
        "scalar": "Nested Scalar Value",
      },
      "scalar": "Relation List Scalar Value",
      "scalarList": Array [
        "Relation List Scalar List Value",
      ],
    },
    Object {
      "id": "id-2",
      "scalar": "Relation List Scalar Value",
      "scalarList": Array [
        "Relation List Scalar List Value",
      ],
    },
  ],
  "scalar": "Scalar Value",
  "scalarList": Array [
    "Scalar List Value",
  ],
}
`;
